ring/ring_test.go

```diff
@@ -1713,105 +1680,188 @@ func TestRing_ShuffleShardWithLookback_CorrectnessWithFuzzy(t *testing.T) {
+				// Replay the events on the timeline.
+				for ix, event := range testData.timeline {
+					switch event.what {
+					case add:
+						ringDesc.Ingesters[event.instanceID] = event.instanceDesc
 
 						ring.ringTokens = ringDesc.GetTokens()
 						ring.ringTokensByZone = ringDesc.getTokensByZone()
 						ring.ringInstanceByToken = ringDesc.getTokensInfo()
 						ring.ringZones = getZones(ringDesc.getTokensByZone())
-					case r < 90:
-						// Scale down instances by 1. To make tests reproducible we get the instance IDs, sort them
-						// and then get a random index (using the random generator initialized with a constant seed).
-						instanceIDs := make([]string, 0, len(ringDesc.Ingesters))
-						for id := range ringDesc.Ingesters {
-							instanceIDs = append(instanceIDs, id)
+						if updateRegisteredTimestampCache {
+							ring.oldestRegisteredTimestamp = ringDesc.getOldestRegisteredTimestamp()
 						}
-
-						sort.Strings(instanceIDs)
-
-						idxToRemove := rand.Intn(len(instanceIDs))
-						idToRemove := instanceIDs[idxToRemove]
-						delete(ringDesc.Ingesters, idToRemove)
+					case remove:
+						delete(ringDesc.Ingesters, event.instanceID)
 
 						ring.ringTokens = ringDesc.GetTokens()
 						ring.ringTokensByZone = ringDesc.getTokensByZone()
 						ring.ringInstanceByToken = ringDesc.getTokensInfo()
 						ring.ringZones = getZones(ringDesc.getTokensByZone())
+						if updateRegisteredTimestampCache {
+							ring.oldestRegisteredTimestamp = ringDesc.getOldestRegisteredTimestamp()
+						}
+					case test:
+						rs, err := ring.ShuffleShardWithLookback(userID, event.shardSize, lookbackPeriod, now).GetAllHealthy(Read)
+						require.NoError(t, err)
+						assert.ElementsMatch(t, event.expected, rs.GetAddresses(), "step %d", ix)
+					}
+				}
+			})
+		}
+	}
+}
 
-						// Remove the terminated instance from the history.
-						for ringTime, ringState := range history {
-							for idx, desc := range ringState.Instances {
-								// In this simulation instance ID == instance address.
-								if desc.Addr != idToRemove {
-									continue
-								}
+func TestRing_ShuffleShardWithLookback_CorrectnessWithFuzzy(t *testing.T) {
+	// The goal of this test is NOT to ensure that the minimum required number of instances
+	// are returned at any given time, BUT at least all required instances are returned.
+	var (
+		numInitialInstances = []int{9, 30, 60, 90}
+		numInitialZones     = []int{1, 3}
+		numEvents           = 100
+		lookbackPeriod      = time.Hour
+		delayBetweenEvents  = 5 * time.Minute // 12 events / hour
+		userID              = "user-1"
+	)
 
-								ringState.Instances = append(ringState.Instances[:idx], ringState.Instances[idx+1:]...)
-								history[ringTime] = ringState
-								break
-							}
-						}
-					default:
-						// Scale up shard size (keeping the per-zone balance).
-						shardSize += numZones
+	for _, updateRegisteredTimestampCache := range []bool{false, true} {
```

Comment:
I would rename `updateRegisteredTimestampCache` to `updateOldestRegisteredTimestamp`.